sourcing/domain/service.go

```diff
@@ -0,0 +1,318 @@
+package domain
+
+import (
+	"context"
+	"math"
+
+	cartDomain "flamingo.me/flamingo-commerce/v3/cart/domain/cart"
+	"flamingo.me/flamingo-commerce/v3/cart/domain/decorator"
+	"flamingo.me/flamingo-commerce/v3/product/domain"
+
+	"flamingo.me/flamingo/v3/framework/flamingo"
+
+	"github.com/pkg/errors"
+)
+
+type (
+	// SourcingService interface
+	SourcingService interface {
+		// AllocateItems returns Sources for the given item in the given cart
+		// e.g. use this during place order to know
+		// throws ErrInsufficientSourceQty if not enough stock is available for the amount of items in the cart
+		// throws ErrNoSourceAvailable if no source is available at all for one of the items
+		// throws ErrNeedMoreDetailsSourceCannotBeDetected if information on the cart (or delivery is missing)
+		AllocateItems(ctx context.Context, decoratedCart *decorator.DecoratedCart) (ItemAllocations, error)
+
+		// GetAvailableSources returns possible Sources for the product and the desired delivery.
+		// Optional the existing cart can be passed so that existing items in the cart can be evaluated also (e.g. deduct stock)
+		// e.g. use this before a product should be placed in the cart to know if and from where an item can be sourced
+		// throws ErrNeedMoreDetailsSourceCannotBeDetected
+		// throws ErrNoSourceAvailable if no source is available for the product and the given delivery
+		GetAvailableSources(ctx context.Context, product domain.BasicProduct, deliveryInfo *cartDomain.DeliveryInfo, decoratedCart *decorator.DecoratedCart) (AvailableSources, error)
+	}
+
+	// ItemID string alias
+	ItemID string
+
+	// ItemAllocations represents the allocated Qtys per itemId
+	ItemAllocations map[ItemID]ItemAllocation
+
+	//ItemAllocation info
+	ItemAllocation struct {
+		AllocatedQtys AllocatedQtys
+		Error         error
+	}
+
+	// AllocatedQtys represents the allocated Qty per source
+	AllocatedQtys map[Source]int
+
+	// Source descriptor for a single location
+	Source struct {
+		// LocationCode identifies the warehouse or stock location
+		LocationCode string
+		// ExternalLocationCode identifies the source location in an external system
+		ExternalLocationCode string
+	}
+
+	// AvailableSources is the result value object containing the available Qty per Source
+	AvailableSources map[Source]int
+
+	//DefaultSourcingService - an example implementation
```

Comment:
It's more than an example, ist't it? It's the main implementation which is used unless you implement something on your own.

Reply:
Updated the doc comment